### PR TITLE
[core] Add an initialize method to the Category and Plugin

### PR DESCRIPTION
**Pull request by jamesonwilliams**
Publish the results of each category's initialization to the Hub.

Related: https://github.com/aws-amplify/amplify-android/issues/209
Related: https://github.com/aws-amplify/amplify-android/issues/215

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amplifyframework/core/Amplify.java [hunk 1]
```diff
@@ -120,13 +127,30 @@ public static void configure(@NonNull final AmplifyConfiguration configuration,
...
+                    return;
+                }
+                Hub.publish(hubChannel, HubEvent.create(
+                    InitializationStatus.FAILED.toString(),
+                    categoryInitializationResult.getPluginInitializationFailures()
+                ));
+            });
+        });
+    }
+
```
Comment (raphkim) [inline core/src/main/java/com/amplifyframework/core/Amplify.java]: <3

File: core/src/main/java/com/amplifyframework/core/InitializationResult.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return InitializationStatus.SUCCEEDED.equals(initializationStatus);
+    }
+
+    /**
+     * Checks if the result is a failure.
+     * @return True if the result is a failure.
+     */
+    public boolean isFailure() {
+        return InitializationStatus.FAILED.equals(initializationStatus);
+    }
```
Comment (raphkim) [inline core/src/main/java/com/amplifyframework/core/InitializationResult.java]: same comment as the one in `CategoryInitializationResult`

File: core/src/main/java/com/amplifyframework/core/plugin/Plugin.java [hunk 3]
```diff
@@ -16,36 +16,67 @@
...
+import com.amplifyframework.core.Amplify;
 import com.amplifyframework.core.category.CategoryTypeable;
 
 import org.json.JSONObject;
 
 /**
- * Interface that defines the contract that every plugin
- * in Amplify System will adhere to.
- * @param <E> The class type of the escape hatch used to circumvent the generic plugin APIs
+ * A plugin is an implementation of a category's behavior. You can implement
```
Comment (raphkim) [inline core/src/main/java/com/amplifyframework/core/plugin/Plugin.java]: > A plugin is an implementation of a category's behavior

Well technically `Plugin` by itself is just an interface... It promises that it's configurable & that it always belongs to a certain category. But to be honest, I'm not sure how better this structure can be documented so I think this is fine

File: core/src/main/java/com/amplifyframework/hub/AWSHubPlugin.java [hunk 4]
```diff
@@ -116,22 +121,25 @@ public void unsubscribe(@NonNull SubscriptionToken subscriptionToken) {
...
     @Override
-    public void configure(@NonNull JSONObject pluginConfiguration, Context context) {
+    public void configure(@NonNull JSONObject pluginConfiguration, @NonNull Context context) {
 
     }
 
+    @NonNull
     @Override
     public Void getEscapeHatch() {
         return null;
```
Comment (raphkim) [inline core/src/main/java/com/amplifyframework/hub/AWSHubPlugin.java]: hmmmmmm

File: core/src/main/java/com/amplifyframework/hub/HubCategory.java [hunk 5]
```diff
@@ -90,4 +98,12 @@ public CategoryType getCategoryType() {
...
         return subscribe(channel, event -> true, transformingListener);
     }
+
+    private HubPlugin<?> getHubPlugin() {
+        if (!super.isConfigured() || super.getPlugins().isEmpty()) {
+            return defaultPlugin;
+        } else {
+            return super.getSelectedPlugin();
+        }
+    }
```
Comment (raphkim) [inline core/src/main/java/com/amplifyframework/hub/HubCategory.java]: nice. I only found out that hub plugins are supposed to be available pre-configuration in that meeting, so I didn't account for that.

File: core/src/main/java/com/amplifyframework/analytics/AnalyticsPlugin.java [hunk 6]
```diff
@@ -25,9 +30,15 @@
  * @param <E> The type of class returned by the plugin implementation to be used as an escape hatch
  */
 public abstract class AnalyticsPlugin<E> implements AnalyticsCategoryBehavior, Plugin<E> {
+    @NonNull
```
[the remark this answers is not in the document]
Comment (raphkim) [reply core/src/main/java/com/amplifyframework/analytics/AnalyticsPlugin.java]: Also applies to a lot of inherited methods from `Category.java` and `Plugin.java`

File: core/src/main/java/com/amplifyframework/hub/AWSHubPlugin.java (hunk 4 above)
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/hub/AWSHubPlugin.java]: Ah yes, this should say `@Nullable`, will update

File: core/src/main/java/com/amplifyframework/core/plugin/Plugin.java (hunk 3 above)
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/core/plugin/Plugin.java]: Eh, yea - just now I tried to make it more technically accurate, I ended up with things like "An implementation of a plugin is an implementation of ..." which is confusing in-and-of-itself.

File: core/src/main/java/com/amplifyframework/core/Amplify.java (hunk 1 above)
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/core/Amplify.java]: Haha

File: aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AmazonPinpointAnalyticsPlugin.java [hunk 7]
```diff
@@ -167,6 +159,7 @@ public void flushEvents() {
     /**
      * {@inheritDoc}
      */
+    @NonNull
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AmazonPinpointAnalyticsPlugin.java]: I don't know the convention around this actually. But Android Studio will suggest to put it on _every_ `@Override` method as well as the base one. Adding this here makes a code inspection warning go away.

File: core/src/main/java/com/amplifyframework/analytics/AnalyticsPlugin.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/analytics/AnalyticsPlugin.java]: (See comment below about Android Studio code inspections.)

File: core/src/main/java/com/amplifyframework/datastore/DataStorePlugin.java [hunk 8]
```diff
@@ -15,12 +15,28 @@
...
+    @NonNull
+    @Override
+    public final CategoryType getCategoryType() {
+        return CategoryType.DATASTORE;
+    }
+
+    @WorkerThread
+    @Override
+    public void initialize(@NonNull Context context) throws AmplifyException {
+    }
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/datastore/DataStorePlugin.java]: Exactly, yea - I wanted to cut down on boiler plate for the actual final plugin implementation. Plugins like our `AndroidLoggingPlugin` don't have any work to do in `intitialize()`, so they don't have to override this method.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java [hunk 9]
```diff
@@ -109,45 +103,46 @@ public void configure(
         } catch (DataStoreException badConfigException) {
             throw new DataStoreException(
                 "There was an issue configuring the plugin from the amplifyconfiguration.json",
-                    badConfigException,
-                    "Check the attached exception for more details and " +
+                badConfigException,
+                "Check the attached exception for more details and " +
                     "be sure you are only calling Amplify.configure once"
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java]: I didn't add `terminate()` as part of this PR, yet, actually :-( That's still needed.

File: core/src/main/java/com/amplifyframework/core/category/CategoryInitializationResult.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        return InitializationStatus.SUCCEEDED.equals(initializationStatus);
+    }
+
+    /**
+     * Checks if the category initialization failed to to a/an error(s).
+     * @return True if category initialization failed
+     */
+    public boolean isFailure() {
+        return InitializationStatus.FAILED.equals(initializationStatus);
+    }
```
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/core/category/CategoryInitializationResult.java]: Yes!! Good catch. Thanks. I'll update.

File: core/src/main/java/com/amplifyframework/core/InitializationResult.java (hunk 2 above)
Comment (jamesonwilliams) [reply core/src/main/java/com/amplifyframework/core/InitializationResult.java]: Updating to the noun pair, success/failure.

**Pull request merged**